### PR TITLE
boards: arm: add Qomu support

### PR DESCRIPTION
**Pull request by michalsieron**
This PR adds support for the Quicklogic Qomu board.
It has the same SoC as QuickFeather, with differences on board level.
Serial is done with USB controller on embedded FPGA.

For more information: [QuickLogic-Corp/qomu-dev-board](https://github.com/QuickLogic-Corp/qomu-dev-board) and [QuickLogic-Corp/qorc-sdk](https://github.com/QuickLogic-Corp/qorc-sdk)

### CODE REVIEW

File: dts/arm/quicklogic/quicklogic_eos_s3.dtsi [hunk 1]
```diff
@@ -49,6 +49,13 @@
 			label = "UART_0";
 		};
 
+		uart1: uart@40020000 {
+			compatible = "quicklogic,usbserial";
+			reg = <0x40020000 DT_SIZE_K(4)>;
+			label = "UART_1";
```
Comment (galak) [inline dts/arm/quicklogic/quicklogic_eos_s3.dtsi:55]: remove label

File: drivers/serial/CMakeLists.txt [hunk 2]
```diff
@@ -32,6 +32,7 @@ zephyr_library_sources_ifdef(CONFIG_UART_STM32 uart_stm32.c)
 zephyr_library_sources_ifdef(CONFIG_UART_SAM0 uart_sam0.c)
 zephyr_library_sources_ifdef(CONFIG_UART_PSOC6 uart_psoc6.c)
 zephyr_library_sources_ifdef(CONFIG_UART_PL011 uart_pl011.c)
+zephyr_library_sources_ifdef(CONFIG_UART_USBSERIAL uart_usbserial.c)
```
Comment (galak) [inline drivers/serial/CMakeLists.txt:35]: rename driver, name is too generic for a driver specific to EOS S3

File: drivers/serial/Kconfig.usbserial [hunk 3]
```diff
@@ -0,0 +1,11 @@
+# QuickFeather USBserialport_S3B configuration option
+
+# Copyright (c) 2022 Antmicro <www.antmicro.com>
+# SPDX-License-Identifier: Apache-2.0
+
+config UART_USBSERIAL
+	bool "USBserialport_S3B serial driver"
+	depends on SOC_EOS_S3
```
Comment (galak) [inline drivers/serial/Kconfig.usbserial:8]: Make this depend on devicetree Kconfg and remove `depends on SOC_EOS_S3`:

```
   default y
   depends on DT_HAS_QUICKLOGIC_USBSERIAL_ENABLED
```

File: boards/arm/qomu/Kconfig.defconfig [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+config BOARD
+	default "qomu"
+
+config UART_PL011
+	default n
+
+config UART_INTERRUPT_DRIVEN
+	default n
+
+config UART_USBSERIAL
```
Comment (galak) [inline boards/arm/qomu/Kconfig.defconfig:17]: remove this, should be determined by devicetree having the node enabled/disabled.

File: boards/arm/qomu/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,5 @@
+# Copyright (c) 2022 Antmicro <www.antmicro.com>
+# SPDX-License-Identifier: Apache-2.0
+
+zephyr_library_sources(board.c)
+zephyr_include_directories(.)
```
Comment (galak) [inline boards/arm/qomu/CMakeLists.txt:5]: remove.  Make board.h include "board.h" instead of <board.h>

**Review by galak — DISMISSED** [review]
a number of changes, to dts, Kconfig, driver, etc.

**Comment by michalsieron** [comment]
Hi @galak, I've addressed all your comments. Could you revisit this PR?

**Comment by fkokosinski** [comment]
Hi @galak, since feature freeze will start soon, could you take a look at this PR again?

**Comment by fkokosinski** [comment]
The one test that is currently failing doesn't seem to be caused by changes introduced in this PR. `frdm_k64f`-related failed builds also seem to affect other PRs.

**Comment by fkokosinski** [comment]
> The one test that is currently failing doesn't seem to be caused by changes introduced in this PR. `frdm_k64f`-related failed builds also seem to affect other PRs.

I've re-run the CI and it looks like it is no longer failing.

**Comment by fkokosinski** [comment]
Hi @dcpleung, as you're assigned to this PR, could you perhaps take a look at it?

**Review by dcpleung — CHANGES_REQUESTED** [review]

File: drivers/serial/Kconfig.eos_s3_usbserial [hunk 6]
```diff
@@ -0,0 +1,12 @@
+# QuickFeather USBserialport_S3B configuration option
+
+# Copyright (c) 2022 Antmicro <www.antmicro.com>
+# SPDX-License-Identifier: Apache-2.0
+
+config UART_EOS_S3_USBSERIAL
+	bool "USBserialport_S3B serial driver"
```
[the remark this answers is not in the document]
Comment (fkokosinski) [reply drivers/serial/Kconfig.eos_s3_usbserial:7]: QuickLogic uses the name "USBserialport_S3B" as the full name of this core e.g. [here](https://github.com/QuickLogic-Corp/s3-gateware/tree/master/usb2serial), so we shouldn't change it IMO.

File: dts/bindings/serial/quicklogic,eos-s3-usbserial.yaml [hunk 7]
```diff
@@ -0,0 +1,15 @@
+# Copyright (c) 2022 Antmicro <www.antmicro.com>
+# SPDX-License-Identifier: Apache-2.0
+
+description: USBserialport_S3B serial interface
```
[the remark this answers is not in the document]
Comment (fkokosinski) [reply dts/bindings/serial/quicklogic,eos-s3-usbserial.yaml:4]: Answered in the other thread.

File: samples/boards/qomu/README.md [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+This sample demonstrates how to load bitstream on EOS-S3 FPGA and use the usbserial driver.
+
+## Requirements
+* Zephyr RTOS with printk enabled
+* [QuickLogic Qomu board](https://github.com/QuickLogic-Corp/qomu-feather-dev-board)
+
+## Building
+
+```bash
+west build -b qomu samples/boards/qomu
```
[the remark this answers is not in the document]
Comment (fkokosinski) [reply samples/boards/qomu/README.md:11]: Done (also changed the format to rst).

File: drivers/serial/Kconfig.eos_s3_usbserial (hunk 6 above)
[the remark this answers is not in the document]
Comment (dcpleung) [reply drivers/serial/Kconfig.eos_s3_usbserial:7]: If that's the case, please add vendor name to the kconfig prompt. Also, please make the kconfig to refer this too (i.e. unify the kconfig and prompt name). Right now, if only looking at the `.config`, it doesn't really corresponding to `USBserialport_S3B`.

File: dts/bindings/serial/quicklogic,eos-s3-usbserial.yaml (hunk 7 above)
[the remark this answers is not in the document]
Comment (dcpleung) [reply dts/bindings/serial/quicklogic,eos-s3-usbserial.yaml:4]: Please add vendor name to the description.

**Review by dcpleung — CHANGES_REQUESTED** [review]

File: drivers/serial/Kconfig.eos_s3_usbserial (hunk 6 above)
[the remark this answers is not in the document]
Comment (fkokosinski) [reply drivers/serial/Kconfig.eos_s3_usbserial]: Done. I've also renamed files that used the previous convention.

File: samples/boards/qomu/README.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (dcpleung) [reply samples/boards/qomu/README.md:11]: Seems like it got reverted back to an early version.

[the remark this answers is not in the document]
Comment (fkokosinski) [reply samples/boards/qomu/README.md]: My bad. Changed it back to the rst version

File: drivers/serial/Kconfig.eos_s3_usbserial (hunk 6 above)
[the remark this answers is not in the document]
Comment (dcpleung) [reply drivers/serial/Kconfig.eos_s3_usbserial:7]: I think this got reverted back to an earlier version. Please add back the vendor name.

File: dts/bindings/serial/quicklogic,eos-s3-usbserial.yaml (hunk 7 above)
[the remark this answers is not in the document]
Comment (dcpleung) [reply dts/bindings/serial/quicklogic,eos-s3-usbserial.yaml:4]: I think this got reverted back to an earlier version. Please add back the vendor name.

**Review by dcpleung — CHANGES_REQUESTED** [review]

File: drivers/serial/Kconfig.eos_s3_usbserial (hunk 6 above)
[the remark this answers is not in the document]
Comment (fkokosinski) [reply drivers/serial/Kconfig.eos_s3_usbserial]: Yes, sorry again - I've accidentally pushed from an older branch. Should be fixed now.

**Review by dcpleung — APPROVED** [review]

**Comment by fkokosinski** [comment]
Hi @galak, since we have 1 approval in this PR and you had requested changes here, could you take a look at this PR again?

**Comment by fkokosinski** [comment]
Seems that @galak is busy. Maybe you @kgugala could take a look at this PR?

**Review by kgugala — APPROVED** [review]
LGTM

**Pull request merged**